File: RRTBase/src/main/java/lab/mars/RRTBase/Vector.java
package lab.mars.RRTBase;

import java.util.Iterator;
import java.util.function.Consumer;

public abstract class Vector<T extends Vector<T>> {

    protected Dimension[] dimensions;

    protected int dimensionCount;

    public Vector(double ... values) {
        dimensions = new Dimension[values.length];
        this.dimensionCount = values.length;
        for (int i = 0; i < values.length;i ++) {
            this.dimensions[i] = new Dimension(values[i]);
        }
    }

    public abstract double distance(T o);

    public abstract double distance2(T o);

    public abstract T normalize();

    public abstract T cpy();

    public abstract double len();

    public abstract double len2();

    public abstract T set(T o);

    public abstract T translate(T v);

    public abstract double dot(T v);

    public abstract T scale(double scalar);

    public abstract T scale(T v);

    public abstract T lerp(T target, double coefficient);

    public abstract boolean equals(Object o);

    public abstract boolean epsilonEquals(T other, double epsilon);

    public abstract String toString();

    public abstract T rotate(double angle);

    public abstract T zero();

    public abstract double angle(T o);

    public abstract T reverse();

    public int dimensionCount() {
        return dimensionCount;
    }

    public Dimension[] dimensions() {
        return dimensions;
    }

    public T forEachDim(Consumer<Dimension> action) {
        for (int i = 0; i < this.dimensions.length; i++) {
            action.accept(dimensions[i]);
        }
        return (T) this;
    }

}

File: MCRRTImp/src/main/kotlin/lab/mars/MCRRTImp/base/Vector.kt
package lab.mars.MCRRTImp.base

abstract class Vector<T : Vector<T>>(vararg values : Double) : Iterable<Dimension>{
    var dimensions : Array<Dimension> protected set

    var dimensionCount : Int protected set

    init {
        dimensionCount = values.size
        dimensions = Array(dimensionCount, { Dimension() })
        values.forEachIndexed { idx, d -> dimensions[idx] assign d }
    }

    abstract fun set(o : T) : T

    abstract fun distance(o : T) : Double

    abstract fun distance2(o : T) : Double

    abstract fun normalize() : T

    abstract fun cpy() : T

    abstract fun len() : Double

    abstract fun len2() : Double

    abstract fun translate(v : T) : T

    abstract fun dot(v : T) : Double

    abstract fun scale(scalar : Double) : T

    abstract fun scale(v : T) : T

    abstract override fun equals(other : Any?) : Boolean

    abstract override fun hashCode() : Int

    abstract override fun toString() : String

    abstract infix fun epsilonEquals(other : T) : Boolean

    abstract fun rotate(angle : Double) : T

    abstract fun zero() : T

    abstract fun angle(other : T) : Double

    abstract fun reverse() : T

    override fun iterator(): Iterator<Dimension> {
        return dimensions.iterator()
    }

    operator fun get(idx : Int) : Double {
        return dimensions[idx].value
    }

    abstract fun crs(other : T): Double
}